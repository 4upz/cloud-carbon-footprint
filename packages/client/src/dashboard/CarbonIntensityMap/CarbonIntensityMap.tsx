/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React, { ReactElement, useState } from 'react'
import {
  Card,
  Box,
  Typography,
  makeStyles,
  Select,
  MenuItem,
  FormControl,
  InputBase,
  createStyles,
  withStyles,
} from '@material-ui/core'
import { ReactComponent as AWSMap } from './AWSMap.svg'
import { ReactComponent as GCPMap } from './GCPMap.svg'
import { ReactComponent as AzureMap } from './AzureMap.svg'

const useStyles = makeStyles(({ palette }) => ({
  root: {
    width: '100%',
    height: '100%',
  },
  title: {
    margin: '0',
    fontSize: '24px',
    fontFamily: 'Helvetica, Arial, sans-serif',
    opacity: '1',
    fontWeight: 900,
    color: 'rgba(0, 0, 0, 0.87)',
    padding: '.2em',
  },
  cloudProvider: {
    color: palette.primary.light,
    textAlign: 'center',
  },
  topContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
    marginBottom: '24px',
  },
  map: {
    width: '100%',
    maxHeight: 700,
  },
}))

const BootstrapInput = withStyles(() =>
  createStyles({
    input: {
      border: '1px solid #ced4da',
      fontSize: 16,
      padding: '10px 26px 10px 12px',
      width: '65px',
      '&:hover': {
        borderColor: 'black',
      },
      '&:focus': {
        backgroundColor: 'white',
        borderRadius: 4,
      },
    },
  }),
)(InputBase)

type CloudProvider = 'AWS' | 'GCP' | 'Azure'

type IntensityMaps = {
  [provider in CloudProvider]: React.ReactNode
}

export const CarbonIntensityMap = (): ReactElement => {
  const [cloudProvider, setCloudProvider] = useState('AWS')
  const classes = useStyles()

  const intensityMaps: IntensityMaps = {
    AWS: <AWSMap className={classes.map} data-testid="awsIntensityMap" />,
    GCP: <GCPMap className={classes.map} data-testid="gcpIntensityMap" />,
    Azure: <AzureMap className={classes.map} data-testid="azureIntensityMap" />,
  }

  const handleChange = (event: React.ChangeEvent<{ value: string }>) => {
    setCloudProvider(event.target.value)
  }

  return (
    <Card className={classes.root}>
      <Box padding={3}>
        <Box className={classes.topContainer}>
          <Typography className={classes.title}>
            Carbon Intensity Map
          </Typography>
          <FormControl variant="outlined">
            <Select
              id="map-select"
              data-testid="select"
              value={cloudProvider}
              onChange={handleChange}
              input={<BootstrapInput />}
            >
              {Object.keys(intensityMaps).map((mapOption) => (
                <MenuItem key={mapOption} id="map-option" value={mapOption}>
                  {mapOption}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </Box>
        <Typography className={`${classes.title} ${classes.cloudProvider}`}>
          {cloudProvider}
        </Typography>
        <Box paddingX={3}>{intensityMaps[cloudProvider]}</Box>
      </Box>
    </Card>
  )
}
