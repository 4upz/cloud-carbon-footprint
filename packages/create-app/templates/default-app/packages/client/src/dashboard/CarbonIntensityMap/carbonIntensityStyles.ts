/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import {
  createStyles,
  InputBase,
  makeStyles,
  withStyles,
} from '@material-ui/core'

export const useStyles = makeStyles(() => ({
  root: {
    width: '100%',
    height: '100%',
  },
  title: {
    margin: '0',
    fontSize: '24px',
    fontFamily: 'Helvetica, Arial, sans-serif',
    opacity: '1',
    fontWeight: 900,
    color: 'rgba(0, 0, 0, 0.87)',
    padding: '.2em',
  },
  topContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
    marginBottom: '24px',
  },
  map: {
    width: '100%',
    maxHeight: 700,
  },
}))

export const BootstrapInput = withStyles(() =>
  createStyles({
    input: {
      border: '1px solid #ced4da',
      fontSize: 16,
      padding: '10px 26px 10px 12px',
      width: '65px',
      '&:hover': {
        borderColor: 'black',
      },
      '&:focus': {
        backgroundColor: 'white',
        borderRadius: 4,
      },
    },
  }),
)(InputBase)
