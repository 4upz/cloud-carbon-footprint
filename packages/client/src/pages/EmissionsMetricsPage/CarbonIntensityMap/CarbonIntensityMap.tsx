/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import React, { ReactElement, useState } from 'react'
import { Card, Box, Typography, Grid } from '@material-ui/core'
import { ReactComponent as AWSMap } from './AWSMap.svg'
import { ReactComponent as GCPMap } from './GCPMap.svg'
import { ReactComponent as AzureMap } from './AzureMap.svg'
import { useStyles } from './carbonIntensityStyles'
import SelectDropdown from '../../../common/SelectDropdown'

type CloudProvider = 'AWS' | 'GCP' | 'Azure'

type IntensityMaps = {
  [provider in CloudProvider]: React.ReactNode
}

const CarbonIntensityMap = (): ReactElement => {
  const [cloudProvider, setCloudProvider] = useState('AWS')
  const classes = useStyles()

  const intensityMaps: IntensityMaps = {
    AWS: <AWSMap className={classes.map} data-testid="awsIntensityMap" />,
    GCP: <GCPMap className={classes.map} data-testid="gcpIntensityMap" />,
    Azure: <AzureMap className={classes.map} data-testid="azureIntensityMap" />,
  }

  const handleChange = (event: React.ChangeEvent<{ value: string }>) => {
    setCloudProvider(event.target.value)
  }

  return (
    <Grid item xs={12}>
      <Card className={classes.root}>
        <Box padding={3}>
          <Box className={classes.topContainer}>
            <Typography className={classes.title}>
              Carbon Intensity Map
            </Typography>
            <SelectDropdown
              id="map"
              value={cloudProvider}
              dropdownOptions={Object.keys(intensityMaps)}
              handleChange={handleChange}
            />
          </Box>
          <Box paddingX={3}>{intensityMaps[cloudProvider]}</Box>
        </Box>
      </Card>
    </Grid>
  )
}

export default CarbonIntensityMap
