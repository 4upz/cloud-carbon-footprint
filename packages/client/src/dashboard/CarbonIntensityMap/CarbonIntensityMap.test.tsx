/*
 * Â© 2021 ThoughtWorks, Inc.
 */

import { CarbonIntensityMap } from './CarbonIntensityMap'
import { fireEvent, render, within } from '@testing-library/react'
import { create } from 'react-test-renderer'

describe('Carbon Intensity Map', () => {
  it('renders with correct configuration', () => {
    const root = create(<CarbonIntensityMap />)

    expect(root.toJSON()).toMatchSnapshot()
  })

  it('renders an svg image with the map for a cloud provider', () => {
    const { getByTestId } = render(<CarbonIntensityMap />)

    const mapImage = getByTestId('intensityMap')
    expect(mapImage).toBeInstanceOf(SVGSVGElement)
  })

  it('renders a dropdown for the cloud providers to choose from', () => {
    const { getByTestId, getByRole } = render(<CarbonIntensityMap />)

    const dropdown = getByTestId('select')
    expect(dropdown).toBeInTheDocument()

    // Clicks on Material UI's select element to display the popup that has the options to be queried
    fireEvent.mouseDown(getByRole('button'))
    const dropdownOptions = within(getByRole('listbox'))

    expect(dropdownOptions.getByText('AWS')).toBeInTheDocument()
    expect(dropdownOptions.getByText('GCP')).toBeInTheDocument()
    expect(dropdownOptions.getByText('Azure')).toBeInTheDocument()
  })
})
