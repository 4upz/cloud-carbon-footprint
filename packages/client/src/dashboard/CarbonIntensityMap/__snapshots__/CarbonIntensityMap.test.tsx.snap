// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carbon Intensity Map should render with correct configuration 1`] = `
<div
  className="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiBox-root MuiBox-root-6"
  >
    <div
      className="MuiBox-root MuiBox-root-7 makeStyles-topContainer-4"
    >
      <p
        className="MuiTypography-root makeStyles-title-2 MuiTypography-body1"
      >
        Carbon Intensity Map
      </p>
      <div
        className="MuiFormControl-root"
      >
        <div
          className="MuiInputBase-root MuiInputBase-formControl"
          data-testid="select"
          onClick={[Function]}
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="map-select"
            className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input WithStyles(ForwardRef(InputBase))-input-8"
            id="map-select"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="button"
            tabIndex={0}
          >
            AWS
          </div>
          <input
            aria-hidden={true}
            className="MuiSelect-nativeInput"
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
            value="AWS"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <p
      className="MuiTypography-root makeStyles-title-2 makeStyles-cloudProvider-3 MuiTypography-body1"
    >
      AWS
    </p>
    <div
      className="MuiBox-root MuiBox-root-9"
    >
      <svg
        className="makeStyles-map-5"
        data-testid="awsIntensityMap"
      >
        AWSMap.svg
      </svg>
    </div>
  </div>
</div>
`;
